src/legacy/shrink.rs

```diff
@@ -0,0 +1,426 @@
+use std::collections::VecDeque;
+use std::io::{self, copy, Read};
+
+use thiserror::Error;
+
+use super::bitstream::{lsb, BitStream};
+
+const MIN_CODE_SIZE: u8 = 9;
+const MAX_CODE_SIZE: u8 = 13;
+
+const MAX_CODE: usize = (1 << MAX_CODE_SIZE) - 1;
+const INVALID_CODE: u16 = u16::MAX;
```

Comment:
Use `Option<u16>` instead. The compiler will pack it into a `u32` on 32-bit and 64-bit platforms, which will save cycles on widening and narrowing instructions.